frontends/concrete-python/examples/sha1/sha1.md

```diff
@@ -0,0 +1,307 @@
+    def xor3(x, y, z):
+        return x ^ y ^ z
+
+    @fhe.function({"x": "encrypted", "y": "encrypted", "z": "encrypted"})
+    def iftern(x, y, z):
+        return z ^ (x & (y ^ z))
+
+    @fhe.function({"x": "encrypted", "y": "encrypted", "z": "encrypted"})
+    def maj(x, y, z):
+        return (x & y) | (z & (x | y))
+
+    @fhe.function({"x": "encrypted"})
+    def rotate30(x):
+        ans = fhe.zeros((32,))
+        ans[30:32] = x[0:2]
+        ans[0:30] = x[2:32]
+        return ans
+
+    @fhe.function({"x": "encrypted"})
+    def rotate5(x):
+        ans = fhe.zeros((32,))
+        ans[5:32] = x[0:27]
+        ans[0:5] = x[27:32]
+        return ans
+
+    @fhe.function({"x": "encrypted", "y": "encrypted"})
+    def add2(x, y):
+        ans = fhe.zeros((32,))
+        cy = 0
+
+        for i in range(32):
+            t = x[i] + y[i] + cy
+            cy, tr = t >= 2, t % 2
+            ans[i] = tr
+
+        return ans
+
+    @fhe.function(
+        {"x": "encrypted", "y": "encrypted", "u": "encrypted", "v": "encrypted", "w": "encrypted"}
+    )
+    def add5(x, y, u, v, w):
+        ans = fhe.zeros((32,))
+        cy = 0
+
+        for i in range(32):
+            t = x[i] + y[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + u[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + v[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + w[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        return ans
+```
+
+It contains 7 functions which can be combined together:
+- `xor3` XORs three values together
+- `iftern` computes the ternary IF, i.e., c ? t : f
+- `maj` computes the majority function
+- `rotate30` rotates a 32-bit word by 30 positions
+- `rotate5` rotates a 32-bit word by 5 positions
+- `add2` adds two values together
+- `add5` adds five values together
+
+This module is then compiled. Remark we set `p_error=10**-8` to a very small value, to avoid
+errors in computations. Thanks to the Module feature, all these functions can be combined
+together, i.e., outputs of some of them can be reused as inputs of some other ones. It is very
+convenient to have a larger function, with some control-flow (conditions, branches, loops) done in
+the clear and operating with these small functions. In our case, it's done in
+`_process_encrypted_chunk_server_side` function.
+
+## Details of `_process_encrypted_chunk_server_side`
+
+`_process_encrypted_chunk_server_side` uses inputs which are encrypted, and returns encrypted
+values. All variables here are 32-bit words when done in the clear. Here, words will be represented
+as 32 1-bit encrypted values, to simplify (and accelerate) the non-linear operations in SHA1.
```

Comment:
> `_process_encrypted_chunk_server_side` uses inputs which are encrypted, and returns encrypted
> values. All variables here are 32-bit words when done in the clear. Here, words will be represented
> as 32 1-bit encrypted values, to simplify (and accelerate) the non-linear operations in SHA1.

Change to:

`_process_encrypted_chunk_server_side` uses encrypted inputs and returns encrypted values. In the clear, all variables are 32-bit words, but here they are represented as 32 1-bit encrypted values to simplify and accelerate the non-linear operations in SHA1.

Reply:
Almost took it. I used

`_process_encrypted_chunk_server_side` uses encrypted inputs and returns encrypted values. In the 
clear, all variables are 32-bit words, but here they are represented as 32 encrypted bits to 
simplify and accelerate the non-linear operations in SHA1.